src/components/MyMap.js

```diff
@@ -0,0 +1,84 @@
+import { Map, Marker, Overlay } from 'pigeon-maps'
+import Tooltip from './Tooltip'
+import { useState } from 'react'
+const haversine = require('haversine')
+
+const osmProvider = (x, y, z) => {
+    const s = String.fromCharCode(97 + ((x + y + z) % 3))
+    return `https://${s}.tile.openstreetmap.org/${z}/${x}/${y}.png`
+}
+
+const MyMap = ({ shops, userLatitude, userLongitude }) => {
+    const [shopTooltipsVisibility, setShopTooltipsVisibility] = useState([])
+    const [userTooltipVisibility, setUserTooltipVisibility] = useState(false)
+
+    const onShopMarkerClick = ({ payload }) => {
+        const index = payload
+        let arrayCopy = [...shopTooltipsVisibility]
+        arrayCopy[index] = !arrayCopy[index]
+        setShopTooltipsVisibility(arrayCopy)
+    }
+
+    const createShopMarkers = (shops) =>
+        shops ? shops.map((shop) =>
+            <Marker
+                key={shop.id}
+                anchor={[parseFloat(shop.x), parseFloat(shop.y)]}
+                color='black'
+                payload={shop.id}
+                onClick={onShopMarkerClick}
+            />
+        ) : null
+
+    const createShopTooltips = (shops) =>
+        shops ? shops.map((shop) =>
+            <Overlay key={shop.id} anchor={[parseFloat(shop.x), parseFloat(shop.y)]}>
+                <Tooltip
+                    isVisible={shopTooltipsVisibility[shop.id]}
+                    name={shop.name}
+                    distance={
+                        Math.round(
+                            haversine({ latitude: userLatitude, longitude: userLongitude },
+                                { latitude: parseFloat(shop.x), longitude: parseFloat(shop.y) }
+                            ) * 10) / 10 + ' km'
+                    } />
+            </Overlay>
+        ) : null
+
+    const createUserMarker = (userLatitude, userLongitude) =>
+        (userLatitude && userLongitude) ?
+            <Marker
+                anchor={[userLatitude, userLongitude]}
+                color='red'
+                payload={'user'}
+                onClick={({ }) => {
+                    setUserTooltipVisibility(!userTooltipVisibility)
+                }}
+            />
+            : null
+
+    const createUserTooltip = (userLatitude, userLongitude) =>
+        (userLatitude && userLongitude) ?
+            <Overlay anchor={[userLatitude, userLongitude]}>
+                <Tooltip
+                    isVisible={userTooltipVisibility}
+                    name={'Your location'}
+                    distance={null} />
+            </Overlay>
+            : null
```

Comment:
I think declaring these functions inside the scope of the main function bloats the main function and makes it more difficult to read.

Reply:
I agree. As it is right now it really looks like a class. Should I just convert it to a class component so it has a `render()` and separate helper methods?

Reply:
I think you can make this much better without switching to class components. But this is up for you to decide how you want to improve things. I don't think there is a rule forbidding function and class components in the same project, but I think consistency is preferable.